## Explain
M1：quality for group A. 
M2：quality for group B. 
M3: waiting time for group A. 
M4: waiting time for group B.
M5: optimal toll for both group A and group B

In [1]:
from gurobipy import *
#set
t=[]# minute
for i in range(0,24*60):
    t.append(i)

In [2]:
#parameter for both group
thope=9*60#desired arrival time
vot=12#value of time
B=6#marginal cost of schedule early
r=9#marginal cost of schedule late
u=20#bottleneck capacity
cost=dict()#schedule cost
for i in t:
    if i < thope:
        cost[i]=B*(thope-i)
    elif i == thope:
        cost[i]=0
    elif i > thope:
        cost[i]=r*(i-thope)

In [3]:
QAB=[300,1200]#total demand
merge=[0.25,0.75]

In [4]:
k=[]
for i in range(len(QAB)):
    k.append(merge[i]/QAB[i])
priority=k.index(max(k))

In [5]:
#variable
m1=Model('groupA_Q')  
m1.setParam( 'OutputFlag', False )
q = m1.addVars(t,lb=0, name = "q" )
priority = 0
m1.update()

#subject to
m1.addConstrs(((q[i]) <= (u*merge[priority]) for i in t),'Capacity')
m1.addConstr(((quicksum(q[i] for i in t)) == QAB[priority] ),'Quality')

#objective
m1.setObjective(quicksum(cost[i]/vot*q[i] for i in t) , GRB.MINIMIZE)
m1.optimize()
#solve
print 'for all q of t minutes whiches value is bigger than zero:'
timeA=[]
for v in m1.getVars():
    if v.x > 0  :
        print("%s: %f" %(v.varName, v.x))
        timeA.append(v.varName)

Academic license - for non-commercial use only
for all q of t minutes whiches value is bigger than zero:
q[505]: 5.000000
q[506]: 5.000000
q[507]: 5.000000
q[508]: 5.000000
q[509]: 5.000000
q[510]: 5.000000
q[511]: 5.000000
q[512]: 5.000000
q[513]: 5.000000
q[514]: 5.000000
q[515]: 5.000000
q[516]: 5.000000
q[517]: 5.000000
q[518]: 5.000000
q[519]: 5.000000
q[520]: 5.000000
q[521]: 5.000000
q[522]: 5.000000
q[523]: 5.000000
q[524]: 5.000000
q[525]: 5.000000
q[526]: 5.000000
q[527]: 5.000000
q[528]: 5.000000
q[529]: 5.000000
q[530]: 5.000000
q[531]: 5.000000
q[532]: 5.000000
q[533]: 5.000000
q[534]: 5.000000
q[535]: 5.000000
q[536]: 5.000000
q[537]: 5.000000
q[538]: 5.000000
q[539]: 5.000000
q[540]: 5.000000
q[541]: 5.000000
q[542]: 5.000000
q[543]: 5.000000
q[544]: 5.000000
q[545]: 5.000000
q[546]: 5.000000
q[547]: 5.000000
q[548]: 5.000000
q[549]: 5.000000
q[550]: 5.000000
q[551]: 5.000000
q[552]: 5.000000
q[553]: 5.000000
q[554]: 5.000000
q[555]: 5.000000
q[556]: 5.000000
q[557]: 5.0

In [6]:
timeforA=[]
for i in timeA:
    timeforA.append(int(i.replace('q[','').replace(']','')))


In [7]:
print timeforA

[505, 506, 507, 508, 509, 510, 511, 512, 513, 514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 526, 527, 528, 529, 530, 531, 532, 533, 534, 535, 536, 537, 538, 539, 540, 541, 542, 543, 544, 545, 546, 547, 548, 549, 550, 551, 552, 553, 554, 555, 556, 557, 558, 559, 560, 561, 562, 563, 564]


In [8]:
#set for time
t1=timeforA
t2=list(set(t).difference(set(timeforA)))
Q=QAB[1]
#variable
m2=Model('numetric')  
m2.setParam( 'OutputFlag', False )
q = m2.addVars(t,lb=0, name = "q" )
m2.update()

#subject to
m2.addConstrs(((q[i]) <= (u*0.75) for i in t1),'Capacity1')
m2.addConstrs(((q[i]) <= (u) for i in t2),'Capacity2')
m2.addConstr(((quicksum(q[i] for i in t)) == Q ),'Quality')

#objective
m2.setObjective(quicksum(cost[i]/vot*q[i] for i in t) , GRB.MINIMIZE)
m2.optimize()
#solve
print 'for all q of t minutes whiches value is bigger than zero:'
for v in m2.getVars():
    if v.x > 0  :
        print("%s: %f" %(v.varName, v.x))

for all q of t minutes whiches value is bigger than zero:
q[495]: 20.000000
q[496]: 20.000000
q[497]: 20.000000
q[498]: 20.000000
q[499]: 20.000000
q[500]: 20.000000
q[501]: 20.000000
q[502]: 20.000000
q[503]: 20.000000
q[504]: 20.000000
q[505]: 15.000000
q[506]: 15.000000
q[507]: 15.000000
q[508]: 15.000000
q[509]: 15.000000
q[510]: 15.000000
q[511]: 15.000000
q[512]: 15.000000
q[513]: 15.000000
q[514]: 15.000000
q[515]: 15.000000
q[516]: 15.000000
q[517]: 15.000000
q[518]: 15.000000
q[519]: 15.000000
q[520]: 15.000000
q[521]: 15.000000
q[522]: 15.000000
q[523]: 15.000000
q[524]: 15.000000
q[525]: 15.000000
q[526]: 15.000000
q[527]: 15.000000
q[528]: 15.000000
q[529]: 15.000000
q[530]: 15.000000
q[531]: 15.000000
q[532]: 15.000000
q[533]: 15.000000
q[534]: 15.000000
q[535]: 15.000000
q[536]: 15.000000
q[537]: 15.000000
q[538]: 15.000000
q[539]: 15.000000
q[540]: 15.000000
q[541]: 15.000000
q[542]: 15.000000
q[543]: 15.000000
q[544]: 15.000000
q[545]: 15.000000
q[546]: 15.000000
q[547]

### (1) At equilibrium, how much is waiting delay of each group? It changes among time of day.
### (2) At equilibrium, how much is equilibrium cost of each group?

In [9]:
#variable
m3=Model('numetric')  
m3.setParam( 'OutputFlag', False )
w = m3.addVars(t,lb=0, name = "w" )
tho=m3.addVar(lb=0,name='tho')
m3.update()
priority=0
#subject to
m3.addConstrs(((-w[i]+tho) <= (cost[i]/vot) for i in t),'q(t)')

#objective
m3.setObjective(-quicksum(u*merge[priority]*w[i] for i in t)+ QAB[priority]*tho , GRB.MAXIMIZE)
m3.optimize()
#solve
print 'equilibrium cost for group A：',m3.getVarByName('tho').x  
print 'for group A waiting time:'
timeA=[] 
for v in m3.getVars():
    if v.x > 0  :
        print("%s: %f" %(v.varName, v.x))
        timeA.append(v.varName)

equilibrium cost for group A： 18.0
for group A waiting time:
w[505]: 1.000000
w[506]: 1.000000
w[507]: 2.000000
w[508]: 2.000000
w[509]: 3.000000
w[510]: 3.000000
w[511]: 4.000000
w[512]: 4.000000
w[513]: 5.000000
w[514]: 5.000000
w[515]: 6.000000
w[516]: 6.000000
w[517]: 7.000000
w[518]: 7.000000
w[519]: 8.000000
w[520]: 8.000000
w[521]: 9.000000
w[522]: 9.000000
w[523]: 10.000000
w[524]: 10.000000
w[525]: 11.000000
w[526]: 11.000000
w[527]: 12.000000
w[528]: 12.000000
w[529]: 13.000000
w[530]: 13.000000
w[531]: 14.000000
w[532]: 14.000000
w[533]: 15.000000
w[534]: 15.000000
w[535]: 16.000000
w[536]: 16.000000
w[537]: 17.000000
w[538]: 17.000000
w[539]: 18.000000
w[540]: 18.000000
w[541]: 18.000000
w[542]: 17.000000
w[543]: 16.000000
w[544]: 15.000000
w[545]: 15.000000
w[546]: 14.000000
w[547]: 13.000000
w[548]: 12.000000
w[549]: 12.000000
w[550]: 11.000000
w[551]: 10.000000
w[552]: 9.000000
w[553]: 9.000000
w[554]: 8.000000
w[555]: 7.000000
w[556]: 6.000000
w[557]: 6.000000
w[558]: 5

In [10]:
#variable
t1=timeforA
t2=list(set(t).difference(set(timeforA)))
priority=1
m4=Model('numetric')  
m4.setParam( 'OutputFlag', False )
w = m4.addVars(t,lb=0, name = "w" )
tho=m4.addVar(lb=0,name='tho')
m4.update()

#subject to
m4.addConstrs(((-w[i]+tho) <= (cost[i]/vot) for i in t),'q(t)')

#objective
m4.setObjective(-quicksum(u*merge[priority]*w[i] for i in t1)-quicksum(u*w[i] for i in t2)+ QAB[priority]*tho , GRB.MAXIMIZE)
m4.optimize()
#solve
print 'equilibrium cost for group B：',m4.getVarByName('tho').x    
print 'for group B waiting time:'
timeA=[]
for v in m4.getVars():
    if v.x > 0  :
        print("%s: %f" %(v.varName, v.x))
        timeA.append(v.varName)

equilibrium cost for group B： 22.0
for group B waiting time:
w[497]: 1.000000
w[498]: 1.000000
w[499]: 2.000000
w[500]: 2.000000
w[501]: 3.000000
w[502]: 3.000000
w[503]: 4.000000
w[504]: 4.000000
w[505]: 5.000000
w[506]: 5.000000
w[507]: 6.000000
w[508]: 6.000000
w[509]: 7.000000
w[510]: 7.000000
w[511]: 8.000000
w[512]: 8.000000
w[513]: 9.000000
w[514]: 9.000000
w[515]: 10.000000
w[516]: 10.000000
w[517]: 11.000000
w[518]: 11.000000
w[519]: 12.000000
w[520]: 12.000000
w[521]: 13.000000
w[522]: 13.000000
w[523]: 14.000000
w[524]: 14.000000
w[525]: 15.000000
w[526]: 15.000000
w[527]: 16.000000
w[528]: 16.000000
w[529]: 17.000000
w[530]: 17.000000
w[531]: 18.000000
w[532]: 18.000000
w[533]: 19.000000
w[534]: 19.000000
w[535]: 20.000000
w[536]: 20.000000
w[537]: 21.000000
w[538]: 21.000000
w[539]: 22.000000
w[540]: 22.000000
w[541]: 22.000000
w[542]: 21.000000
w[543]: 20.000000
w[544]: 19.000000
w[545]: 19.000000
w[546]: 18.000000
w[547]: 17.000000
w[548]: 16.000000
w[549]: 16.000000
w[5

the equilibrium cost for A group is 18 and for B group is 22

### (3)At equilibrium, how much is total trip cost of social?

In [11]:
TCA=0
for i,j in enumerate(m1.getVars()):
    TCA+=m1.getVars()[i].x*(cost[i]+vot*m3.getVars()[i].x)
print 'total trip cost for group A: ',TCA
TCB=0
for i,j in enumerate(m2.getVars()):
    TCB+=m2.getVars()[i].x*(cost[i]+vot*m4.getVars()[i].x)
print 'total trip cost for group B: ',TCB

total trip cost for group A:  65880.0
total trip cost for group B:  321180.0


### (4)To achieve optimal situation, how much is the time-varying optimal toll?

In [12]:
#variable
m5=Model('numetric')  
m5.setParam( 'OutputFlag', False )
p = m5.addVars(t,lb=0, name = "p" )
tho=m5.addVar(lb=0,name='tho')
m5.update()
priority=0

#subject to
m5.addConstrs(((-p[i]+tho*vot) <= (cost[i]) for i in t),'q(t)')

#objective
m5.setObjective(-quicksum(u*p[i] for i in t)+ (QAB[0]+QAB[1])*tho*vot , GRB.MAXIMIZE)
m5.optimize()
#solve
print ' toll for group A :'
for v in m5.getVars():
    if v.x > 0  :
        print("%s: %f" %(v.varName, v.x))

 toll for group A :
p[496]: 6.000000
p[497]: 12.000000
p[498]: 18.000000
p[499]: 24.000000
p[500]: 30.000000
p[501]: 36.000000
p[502]: 42.000000
p[503]: 48.000000
p[504]: 54.000000
p[505]: 60.000000
p[506]: 66.000000
p[507]: 72.000000
p[508]: 78.000000
p[509]: 84.000000
p[510]: 90.000000
p[511]: 96.000000
p[512]: 102.000000
p[513]: 108.000000
p[514]: 114.000000
p[515]: 120.000000
p[516]: 126.000000
p[517]: 132.000000
p[518]: 138.000000
p[519]: 144.000000
p[520]: 150.000000
p[521]: 156.000000
p[522]: 162.000000
p[523]: 168.000000
p[524]: 174.000000
p[525]: 180.000000
p[526]: 186.000000
p[527]: 192.000000
p[528]: 198.000000
p[529]: 204.000000
p[530]: 210.000000
p[531]: 216.000000
p[532]: 222.000000
p[533]: 228.000000
p[534]: 234.000000
p[535]: 240.000000
p[536]: 246.000000
p[537]: 252.000000
p[538]: 258.000000
p[539]: 264.000000
p[540]: 270.000000
p[541]: 261.000000
p[542]: 252.000000
p[543]: 243.000000
p[544]: 234.000000
p[545]: 225.000000
p[546]: 216.000000
p[547]: 207.000000
p[548]: 1

### (5) Under the optimal time-varying toll, how much is equilibrium cost of each group?

In [13]:
print 'equilibrium cost for group A：',m5.getVarByName('tho').x    
print 'equilibrium cost for group B：',m5.getVarByName('tho').x   

equilibrium cost for group A： 22.5
equilibrium cost for group B： 22.5


### (6) Under the optimal time-varying toll, how much is the total revenues (toll payments)?

In [14]:
TotalRevenue=0
for i,j in enumerate(m1.getVars()):
    TotalRevenue+=m1.getVars()[i].x*m5.getVars()[i].x
for i,j in enumerate(m2.getVars()):
    TotalRevenue+=m2.getVars()[i].x*m5.getVars()[i].x
print 'total revenue: ',TotalRevenue

total revenue:  202500.0


### (7)Compare the equilibrium cost of each group between cases with and without the optimal toll charged. Everyone happy?

In [15]:
print 'equilibrium cost for group A without toll：',m3.getVarByName('tho').x 
print 'equilibrium cost for group B without toll：',m4.getVarByName('tho').x 
print 'equilibrium cost under toll：',m5.getVarByName('tho').x 

equilibrium cost for group A without toll： 18.0
equilibrium cost for group B without toll： 22.0
equilibrium cost under toll： 22.5


### conclusion for (7): 
Comparing the equilibrium cost of each group between cases with and without the optimal toll charged, the equilibrium cost in both group A and B are less than the equilibrium cost under toll. As the result, both group A and B are not happy.

### (8)Under the optimal time-varying toll, when the revenues are refunded for everyone equally, how much is the cost for each group? Everyone happy?

In [16]:
refund=float(TotalRevenue)/(QAB[0]+QAB[1])
print refund

135.0


In [17]:
changeA=refund*QAB[0]-TCA
changeB=refund*QAB[1]-TCB

In [20]:
print changeA,changeB

-25380.0 -159180.0


In [19]:
NewTotalRevenue = -(changeA+changeB)
print NewTotalRevenue

184560.0


### conclusion for (8): 
The cost for group A is 25380. And the cost for group B is 159180.
The new total cost is less than the old one.
Both are less than the total cost without refund. So, I think everyone is happy.